### PR TITLE
refactor: cmd and plugin engine modules, and fixing some bugs

### LINKED ISSUE #64: Refactor: cmd and pluginengine Refactor
## Description

Current situation:

- most logic (download plugins, create a backend, read state, create a plan and execute it) all lives in the `cmd/devstream/install.go`; the `internal/pkg/pluginengine` doesn't do much except load the so and execute the corresponding interface
- plan generation code lives in `stream/internal/pkg/planmanager/result.go` which doesn't seem right according to the name
- plan generation calls the engine instead of the other way around

## Describe the Proposal

- move the main logic out of the `cmd/devstream` package and into the `internal/pkg/pluginengine` package
- pluginengine as the main entrance (as shown in the architecture diagram)
- figure out a way so that plan doesn't call engine, but the other way around, so that:
- pluginengine initiates config loading, plugin downloading, plan creating/executing, and state managing

### LINKED ISSUE #102: init and install logic improvement:
What should happen when install is called without calling init first?

Shall we run the init anyway for the user, or only a warning should be given and let the user take control of the whole process?

We can refer to other open-source tools, like Terraform init/apply.

### PR DESCRIPTION
**Pull request by IronCore864**
# Summary

## Key Points

- [x] This is a breaking change.

## Description

Build:
- `make build` will make sure the default plugins directory exists (`.devstream`) and will build the plugins to `.devstream/*.so`
- `make fmt` changed to `go install` (starting in Go 1.17, installing executables with go get is deprecated)

Command/Engine:
- init: the logic isn't changed; apply without init will also work.
- apply: the `applyCMDFunc` is hugely reduced, moving the main logic to a new function `pluginengine.Run`
- a new helper file for the engine module is created to better organize the code structure

Configloader:
- viper initialization is moved out of the `initConfig` function (bug: `dtm` without any subcommand would fail because it tries to load the config file)
- viper UnmarshalKey is moved to `configloader.LoadConf` to reduce duplicated code

State/Backend:
- default state file is changed from `.devstream/devstream.state` to `devstream.state`
- local backend will create a new empty file if it doesn't exist. This solves the minor bug that the first time when dtm is executed, there is no state, hence stateMap reading would cause an error output
- fixing a bug that after installation, the state is still "running" instead of "installed"

Plugins:
- default plugins directory: changed from `.devstream/plugins` to `.devstream`.
- argocd plugin: cascading deletion of an app when the Application CRD is deleted (for easier local testing)

**Note: the changes of state/plugins are due to the fact that if using `-p` to specify another directory for plugins, the state file will still live at `.devstream/devstreamstate` but the `.devstream` folder might not exist since it's not created by the `init`.**

Plan:
- `plan.Execute` now is moved to the plugin engine module, so that there are no circular references (engine uses plan.execute, which uses engine.install). Because of this, `change.Action` is removed; only `change.ActionName` is used.
- a `plan_helper` is created to better organize the code, and the `result` is separated with `plan`

Logging:
- removed "===" prefixes and suffixes (will be further improved by introducing another logging module)
- fixed some grammar errors

Test:
- fixed a commented test

### Related Issues

Closes #64 
Closes #102

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,4 @@ dtm
 
 # .devstream
 .devstream/
+devstream.state
```
Comment (daniel-hutao) [inline .gitignore:30]: `devstream.state` will located inside the `.devstream/` directory, so I think it's no need to put the `devstream.state` here.

File: internal/pkg/backend/local/local.go [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
+	"errors"
 	"io/ioutil"
+	"log"
 	"os"
 	"path/filepath"
 	"sync"
 )
 
-const DefaultStateFile = ".devstream/devstream.state"
+const DefaultStateFile = "devstream.state"
```
Comment (daniel-hutao) [inline internal/pkg/backend/local/local.go:12]: Our previous plan was to put this file in the `.devstream/` directory, why did we decide to change it?

File: internal/pkg/backend/local/local.go [hunk 3]
```diff
@@ -21,6 +23,15 @@ func NewLocal(filename string) *Local {
 	if filename == "" {
 		lFile = DefaultStateFile
 	}
+
+	if _, err := os.Stat(lFile); errors.Is(err, os.ErrNotExist) {
```
Comment (daniel-hutao) [inline internal/pkg/backend/local/local.go:27]: `Write()` will create the state file if it's not exist

File: internal/pkg/planmanager/result.go [hunk 4]
```diff
@@ -31,67 +29,17 @@ func (p *Plan) handleResult(change *Change) error {
 	// uninstall failed
 	if change.ActionName == statemanager.ActionUninstall && !change.Result.Succeeded {
 		state.Status = statemanager.StatusInstalled
-		log.Printf("=== plugin %s uninstall failed ===", change.Tool.Name)
-		// install or reinstall failed
+		log.Printf("Plugin %s uninstall failed.", change.Tool.Name)
```
Comment (daniel-hutao) [inline internal/pkg/planmanager/result.go:32]: Log printing generally starts with lowercase

File: internal/pkg/configloader/config.go [hunk 5]
```diff
@@ -28,6 +31,31 @@ func (t *Tool) DeepCopy() *Tool {
 	return &retTool
 }
 
+// LoadConf reads an input file as a Config struct.
+func LoadConf(fname string) *Config {
```
Comment (daniel-hutao) [inline internal/pkg/configloader/config.go:35]: These codes are duplicated in `cmd/main.go`

**Review by daniel-hutao — CHANGES_REQUESTED** [review]
This pr is a bit too big, it looks a little hard to understand. That's all from me.

File: .gitignore (hunk 1 above)
Comment (IronCore864) [reply .gitignore:30]: `devstream.state` now is moved out of the `.devstream/` folder.

File: internal/pkg/backend/local/local.go (hunk 2 above)
Comment (IronCore864) [reply internal/pkg/backend/local/local.go:12]: See the description for the details.

TL;DR: when specifying a plugin directory with "-p", this folder won't be created so there is an error when creating the state file.

File: internal/pkg/planmanager/result.go (hunk 4 above)
Comment (IronCore864) [reply internal/pkg/planmanager/result.go:32]: At the moment, we have different formats, which will be improved in the next sprint. Closing this comment for now since we will improve it in the next sprint thoroughly.

Some discussion:

- Log messages are line-oriented, so they may be full sentences. For full sentences, the first letter should be uppercase and it should end with punctuation.
- Au-contraire, errors should start with lowercase letters.

File: internal/pkg/configloader/config.go (hunk 5 above)
Comment (IronCore864) [reply internal/pkg/configloader/config.go:35]: cmd.main sets viper.AutomaticEnv() and binds ENV vars and flags (cmd related);

LoadConf reads the config file (config file related, hence the name "load conf")

**Comment by IronCore864** [comment]
Added some fixes based on code review:

- [ ] state: needs more investigation;
- [x] pluginengine.Do should return an error;
- [x] internal/pkg/planmanager/change.go has duplicated code and comments.

**Review by daniel-hutao — APPROVED** [review]

**Review by lfbdev — APPROVED** [review]
/lgtm
